167 41  connect with api profile chips

NadezhdaSakal opened the pull request.

NadezhdaSakal committed:
connect with api for chips

NadezhdaSakal committed:
Merge remote-tracking branch 'origin/develop' into 167-41--connect-with-api-profile-chips

NadezhdaSakal committed:
add ChipsInteractor

NadezhdaSakal committed:
fix header

NadezhdaSakal committed:
add recyclers for chips

NadezhdaSakal committed:
Merge remote-tracking branch 'origin/develop' into 167-41--connect-with-api-profile-chips

# Conflicts:
#	app/src/main/java/com/innoprog/android/feature/profile/profiledetails/presentation/ProfileViewModel.kt

NadezhdaSakal committed:
connect with api chip All

NadezhdaSakal committed:
fix detekt

NadezhdaSakal committed:
fix connect with api chip All

NadezhdaSakal committed:
fix chip project

NadezhdaSakal committed:
Merge remote-tracking branch 'origin/develop' into 167-41--connect-with-api-profile-chips

# Conflicts:
#	app/src/main/java/com/innoprog/android/network/data/NetworkModule.kt

NadezhdaSakal committed:
delete AuthInterceptor & add if

NadezhdaSakal committed:
fix after detekt

NadezhdaSakal committed:
fix layout

NadezhdaSakal committed:
fix connect chips

NadezhdaSakal committed:
fix actions

AntaOtk reviewed (commented):
вопрос по работе.
почему на проектах передаем запрос  https://bbahtb5q32tp628tg82c.containers.yandexcloud.net/v1/feed?type=NEW?

NadezhdaSakal commented:
Я предположила, что в фигме ошибка, так как у проекта нет полей ни про комменты, ни про лайки. Также, судя по чипу Всё, мы запрашиваем NEWS и IDEA, логично предположить, что по соответствующим чипам идут запросы news, idea.

NadezhdaSakal committed:
Merge remote-tracking branch 'origin/develop' into 167-41--connect-with-api-profile-chips

# Conflicts:
#	app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/impl/ProfileInfoRepoImpl.kt
#	config/detekt/detekt.yml

NadezhdaSakal committed:
resolve conflicts

NadezhdaSakal committed:
fix after detekt

Commit:
Merge branch 'develop' into 167-41--connect-with-api-profile-chips

# Conflicts:
#	app/src/main/java/com/innoprog/android/network/data/ApiConstants.kt

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/impl/ChipsProfileRepoImpl.kt:
Все методы повторяют одну и ту же логику, лучше вынести общую логику в один метод, пример:
```kotlin
private inline fun <reified Data, reified Domain> getResult(
        crossinline getResponse: suspend () -> Response,
        crossinline mapToDomain: (Data) -> Domain
    ): Flow<Resource<Domain>> {
        return flow {
            val response = getResponse()
            if (response is Data && response.resultCode == ApiConstants.SUCCESS_CODE) {
                emit(Resource.Success(mapToDomain(response)))
            } else {
                emit(Resource.Error(getErrorType(response.resultCode)))
            }
        }
    }
```

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/impl/ProfileInfoRepoImpl.kt:
Лучше вынести мапперы в extension функции, по аналогии с FeedDto

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/network/Request.kt:
Именно Data класс тут не обязательно использовать как и в ChipsScreenState, он генерит методы которые мы не используем что замедляет релизную сборку + лучше более явно задавать имя для sealed interface, чтобы было понятно к какой логике он относится. Например RequestByProfile

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/network/RetrofitClient.kt:
```diff
@@ -0,0 +1,65 @@
+    companion object {
+
+        private const val IDEA = "IDEA"
+        private const val NEWS = "NEWS"
```
Из develop пришел enum class Publication type, лучще использовать его для единообразности
```kotlin
ChipsResponse(service.getIdeas(type = PublicationType.IDEA.value, authorId = dto.authorId))
```

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/data/network/RetrofitClient.kt:
```diff
@@ -0,0 +1,65 @@
+                        throw IllegalArgumentException("Unsupported request type")
+                    }
+                }
+                response.apply { resultCode = ApiConstants.SUCCESS_CODE }
```
Необходимо добавить присвоение кода ответа от сервера, для их дальнейшей обработки. Для этого в ProfileApi нужно обернуть каждый возвращаемый тип в Response<>
```kotlin
@GET("/v1/profile")
suspend fun loadProfile(): Response<ProfileResponse>
```

NadezhdaSakal replied:
оборачиваем в наш респонз или рертрофит?

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/presentation/adapter/PublicationsViewHolder.kt:
```diff
@@ -0,0 +1,81 @@
+                        .into(ivProjectLogo)
+
+                    tvProjectName.text = "News Project"
+                    tvProjectDirection.text = "News Direction"
```
Остались вручную прописанные значения, думаю тут нужно поменять на то что приходит в модельке

NadezhdaSakal replied:
не совсем понимаю что у нас в модельке есть  tvProjectName и tvProjectDirection

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/presentation/adapter/PublicationsViewHolder.kt:
Лучше разбить логику на несколько методов + вынести общую. Например это используется и при идее и при новости.
```kotlin
tvPublicationTitle.text = publication.title
tvPublicationContent.text = publication.content
tvPublicationAuthorName.text = publication.author.name
tvCommentsCount.text = publication.commentsCount.toString()
tvLikesCount.text = publication.likesCount.toString()
```
Еще была информация от заказчика что аватарки пользователя нет и мы в этот кружок помещаем инициалы имени и фамилии. Это уже реализовано на экране списка курсов, можно взять оттуда

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/presentation/adapter/PublicationsRecyclerAdapter.kt:
```diff
@@ -0,0 +1,29 @@
+        holder.bind(publications[position])
+        holder.itemView.setOnClickListener {
+            onPublicationClick.invoke(publications[position])
+        }
```
Лучше делать установку clickListener в методе onCreateViewHolder, чтобы эта логика выполнялась один раз при создании ViewHolder

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/profile/profiledetails/presentation/ProfileViewModel.kt:
Лучше вынести общую логику методов в отдельный метод:
```kotlin
private inline fun <reified D> runSafelyUseCase(
        crossinline getUseCaseFlow: suspend () -> Flow<Resource<D>>,
        crossinline onSuccess: (D) -> Unit
    ) {
        viewModelScope.launch(Dispatchers.IO) {
            runCatching {
                getUseCaseFlow().collect { result ->
                    when (result) {
                        is Resource.Success -> onSuccess(result.data)
                        is Resource.Error -> {
                            _chipsUiState.postValue(ChipsScreenState.Error(result.errorType))
                        }
                    }
                }
            }.onFailure { error ->
                if (BuildConfig.DEBUG) {
                    Log.v(TAG, "error -> ${error.localizedMessage}")
                    error.printStackTrace()
                }
                _chipsUiState.postValue(ChipsScreenState.Error(type = ErrorType.NO_CONNECTION))
            }
        }
    }
```
Тогда другие методы сократятся до такого объема:
```kotlin
fun loadChipLiked(pageSize: Int) {
        runSafelyUseCase<List<FeedWrapper>>(
            getUseCaseFlow = { chipsInteractor.getLikes(pageSize) }
        ) { _chipsUiState.postValue(ChipsScreenState.Liked(it)) }
    }
```

NadezhdaSakal replied:
ругается на BuildConfig.DEBUG

Commit:
Add examples of common methods, fix item_news.xml guideline distance, move setOnClickListener to onCreateViewHolder method, resolve merge conflict

Secur1ty-21 commented on app/src/main/res/layout/fragment_profile.xml:
```diff
@@ -123,25 +123,77 @@
 
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+            <androidx.recyclerview.widget.RecyclerView
+                android:id="@+id/recycler_all"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:padding="@dimen/padding_8"
+                android:visibility="gone"
+                app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
+                tools:itemCount="2"
+                tools:listitem="@layout/item_news" />
+
+            <androidx.recyclerview.widget.RecyclerView
+                android:id="@+id/recycler_projects"
```
Думаю можно обойтись одним RecyclerView

NadezhdaSakal replied:
ок

Commit:
Return detekt rules, refactoring, add marker NetworkModel for detekt

Secur1ty-21 reviewed (commented):
После внесения исправлений по адаптеру и убиранию лишних recyclerView, ошибок detekt быть не должно

NadezhdaSakal committed:
add fun getResult & refactor
delete recyclers & refactor fragment
add fun runSafelyUseCase & refactor
rename Request

NadezhdaSakal committed:
refactor sealed interfaces

NadezhdaSakal committed:
add extensions & refactor ProfileInfoRepoImplRepo

NadezhdaSakal committed:
add Response wrapper

NadezhdaSakal committed:
add catch blocks
refactor viewHolder

NadezhdaSakal committed:
after detekt

NadezhdaSakal committed:
Merge branch 'develop' into 167-41--connect-with-api-profile-chips

NadezhdaSakal committed:
refactor avatar profile

NadezhdaSakal committed:
Merge remote-tracking branch 'origin/167-41--connect-with-api-profile-chips' into 167-41--connect-with-api-profile-chips

NadezhdaSakal committed:
fix client

NadezhdaSakal committed:
fix client

Commit:
Fix news view, fix get response in ProfileRetrofitClient, fix show placeholder logic, refactoring

NadezhdaSakal committed:
add project

Commit:
Add get project by id request, add feedWithProject model

NadezhdaSakal committed:
bind FeedWithProject

NadezhdaSakal committed:
bind FeedWithProject at ui

NadezhdaSakal committed:
fix after detekt
